generate/avro.go

```diff
@@ -0,0 +1,304 @@
+package generate
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+const (
+	avroHeader           = `{"name":%q,"type":"record","fields":[{"name":"AvroWriteTime","doc":"The timestamp when this avro data is written. Useful for identifying the newest row of data sharing keys.","type":"long","logicalType":"timestamp-millis"},{"name":"AvroDeleted","doc":"This is set to true when the Avro data is recording a delete in the source data.","default":false,"type":"boolean"},`
+	embeddedStructHeader = `{"name":%q,"type":"record","fields":[`
+)
+
+type avroConfig struct {
+	dir       string
+	namespace string
+	writer    io.Writer
+}
+
+// buildAvroSerializationFunctions will run gogen-avro on the given schema to build a serialization and deserialization
+// code for the schema.
+func buildAvroSerializationFunctions(outputDir string) error {
+	// TODO buildAvroSerializationFunctions - this should take the schema and run gogen-avro
+	return nil
+}
+
+// buildAvroSchemaFile will generate an Avro schema based on the go struct with the given name in the file at the
+// given path. I will write it to the same directory as the go file as `<name>.avsc` but with the name lowercased.
+// http://avro.apache.org/docs/current/spec.html
+//
+// By default the file is written with now whitespace to minimize the size, choose pretty for better formatting
+//
+// Note: Avro can't handle maps with a key other than a string, http://avro.apache.org/docs/current/spec.html#Maps
+// Neither can JSON schema, https://json-schema.org/understanding-json-schema/reference/object.html so this only
+// becomes relevant if it is used with go structs which weren't just generated from JSON schema
+func buildAvroSchemaFile(name, dir string, spec *ast.TypeSpec, pretty bool) (string, error) {
```

Comment:
I find it interesting that we're building the Avro schema based off of the Go structs and not the JSON Schema but I did see your comment about why that decision was made and I'm happy with this solution too